src/orange/OrangeTrackView.hh

```diff
@@ -1209,12 +1200,63 @@ CELER_FUNCTION TransformId OrangeTrackView::get_transform(DaughterId daughter_id
 /*!
  * Get the transform ID for the given daughter.
  */
-CELER_FUNCTION TransformId OrangeTrackView::get_transform(LevelId lev)
+CELER_FUNCTION TransformId OrangeTrackView::get_transform(LevelId lev) const
 {
     CELER_EXPECT(lev < this->level());
     LSA lsa(&states_, track_slot_, lev);
     return this->get_transform(this->get_daughter(lsa));
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Get the normal vector of the current surface as defined by the geometry.
+ */
+CELER_FUNCTION Real3 OrangeTrackView::geo_normal() const
+{
+    CELER_EXPECT(this->is_on_boundary());
+
+    auto lsa = this->make_lsa(this->surface_level());
+    TrackerVisitor visit_tracker{params_};
+    auto normal = visit_tracker(
+        [pos = lsa.pos(), local_surface = this->surf()](auto&& t) {
+            return t.normal(pos, local_surface);
+        },
+        lsa.universe());
+
+    // Rotate normal up to global coordinates
+    auto apply_transform = TransformVisitor{params_};
+    auto rotate_up = [&normal](auto&& t) { normal = t.rotate_up(normal); };
+    for (auto level : range<int>(this->level().unchecked_get()).step(-1))
+    {
+        apply_transform(rotate_up, this->get_transform(LevelId(level)));
+    }
+
+    CELER_ENSURE(is_soft_unit_vector(normal));
+
+    return normal;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the normal vector of the current surface.
+ *
+ * When called after \c cross_boundary this function gives the surface
+ * normal pointed into the previous volume. This ensures that the dot product
+ * of the surface normal and the track direction is negative.
```

Comment:
This is true only if it hasn't changed direction while on the boundary; it could "outside pointing in" and still have the same normal. I think your github comment phrased it better.

I'd add to the class documentation up top (since this function is hard to find): Even if we're numerically and logically 'on' the surface and the direction is headed away from it, the normal is toward the surface *no matter what*  based on the stored sense.